lib/vrl/compiler/src/function.rs

```diff
@@ -25,7 +26,7 @@ pub type Compiled = Result<Box<dyn Expression>, Box<dyn DiagnosticMessage>>;
 pub type CompiledArgument =
     Result<Option<Box<dyn std::any::Any + Send + Sync>>, Box<dyn DiagnosticMessage>>;
 
-pub trait Function: Send + Sync + fmt::Debug {
+pub trait Function: Send + Sync + fmt::Debug + DynClone {
```

Comment:
This does feel a little wasteful. I'm wondering if instead of cloning the functions for each VM we could make the VM hold:

```rust
    fns: Arc<Vec<Box<dyn Function>>>,
```

Is there any reason why that wouldn't work?

Reply:
We could, I don't think there's a reason that couldn't work. At the same time, `Function` is only used at compile-time, and if you look at all the stdlib functions that implement `Function`, all of them are stateless (and most/all implement `Copy`).

So, I _highly_ doubt you'll see any difference in actual performance.

Happy to make the change, though, if you'd like for it to be wrapped in an `Arc`.

Reply:
Yeah it's not so much speed I'm thinking of, rather just using up a few extra bytes (maybe even kilobytes!) of ram that we possibly don't need to use.  If it's not an easy change then I don't think it will be a problem leaving it like this.

Reply:
I don't think there'll be any impact, but it was an easy change to make 👍

[bd842fb](https://github.com/vectordotdev/vector/pull/12549/commits/bd842fb22b9e03fd0373cf9d5a2da7c49bed87cd)

Reply:
I don't think there'll be any impact, but it was an easy change to make 👍

[bd842fb](https://github.com/vectordotdev/vector/pull/12549/commits/bd842fb22b9e03fd0373cf9d5a2da7c49bed87cd)